### DISCUSSION TITLE
Moving particle to deeper depth based on ROMS salinity

### DISCUSSION THREAD

**Discussion by jwongala** [Parcels-code/Parcels#1503] [Q&A]
Hello! I am investigating the best way to move particles to depth based on salinity in the fieldset. I found a few options: using the Argo float example, particle field interaction, or a combination of the two. I think the argo float example might be the way to go, because I don't want to store a new variable in the fieldset, just move the particle to different depths. If anyone has any suggestions on which option would be the best, and other other suggestions, I would really appreciate it.

**Comment by jwongala** [comment]
I am also using a nested fields set up with two ROMS.

**Comment by erikvansebille** [comment]
OK, so do I understand correctly that you want to place the particle at a certain depth based on its salinity? 

Perhaps you can find some relevant ideas in what @daanreijnders did in his paper at https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2021MS002850?

**Reply by jwongala** [reply]
Yes that is correct. I will read this paper thank you!

**Comment by daanreijnders** [comment]
I haven't encountered a problem like that before, but I think you could write a kernel for that does a salinity search based on bisection search. It could roughly work as follows:
 - Define a depth interval in which you think the target salinity lies: [z_min, z_max]. 
 - Sample the salinities at the edges of the interval and subtract the target salinity (so you get the 'anomalies' with respect to it)
 - If there is a sign change between the edges of the interval, your target salinity lies somewhere in between, so you can continue 🎉. If not, it probably is outside your bounds, as salinity profiles are pretty monotonic.
 - Find the depth in the center of your interval (z_mid = z_min + (zmax-zmin)/2). Use this to create two intervals: [z_min, z_mid] and [z_mid, z_max]. You can then iteratively apply this procedure to find the target salinity to arbitrary precision, and put the particle there. 
 
 If you use JIT-mode (which interpolates a lot faster than Scipy-mode), I don't think you can work with loops (not sure anymore, but you can test this). But if you use [status codes ](https://docs.oceanparcels.org/en/latest/examples/tutorial_kernelloop.html#Working-with-Status-Codes)(repeat = 20) and cleverly store some variables with the particle, you can let the kernel repeat the time step, effectively creating a loop.

Hope this helps and I'm curious about the outcome :)

**Reply by jwongala** [reply]
Thank you Daan this is very helpful. I have a few questions. 

- Do you think I could change the order of this assign a target salinity, or range of salinities, then sample the salinity at the particles location, and if the salinity is within that range, move it to that depth layer? The purpose of modeling this behavior is because the parasite I am modeling is known to floculate in seawater, then sink to depth, which is why I think it would be appropriate to change the order. 
- What do you mean by the edges of the interval? Apologies my python knowledge is limited to using parcels.

**Reply by daanreijnders** [reply]
Yes, I don’t think the order would be an issue.

I mean the edges of an interval in a mathematical sense, so z_min and z_max. Let me know if the other suggestions make sense. I don’t have time to write a kernel for you, but happy to help with some tips.

**Reply by daanreijnders** [reply]
Also, [this might actually be a better explanation](https://en.m.wikipedia.org/wiki/Bisection_method) of the concept than my own.

**Reply by jwongala** [reply]
Thanks I appreciate it. And no worries, I will work on it and message here if I have any questions.
